Feat/vacancy duplicate check

edo9k opened the pull request:
Bloqueia a duplicação de vaga corretamente nos casos de Local e Restrição repetidos, e também nos casos de de ambos, ou um dos dois campos estar nulo.

edo9k committed:
Added duplicate check

Still needs more extensive testing to check if the complex
null/not-null logic is working.

edo9k committed:
Corrected duplication check

Correcly blocking duplicates with same Region/Resctriction, with null
Region, null Restriction and with both optional properties being null.

annecchini commented on app/validators/vacancy.js:
```diff
@@ -89,6 +99,39 @@ const validateReserve = (value, db, mode, item) => {
+
+    const restriction_id = (typeof body !== 'undefined' && body.restriction_id) 
+      || (typeof item !== 'undefined' && item.restriction_id)
+      || null
```
O que acontece quando eu tento mudar (update) region_id de "valorAtual" para null?
R: Ele vai tentar fazer a validação com o valor de item.region_id correto? Mas isso está errado. Ele tem que tentar fazer a validação com null (O novo valor de region_id).

O que acontece quando eu tento mudar (update) restriction_id de "valorAtual" para null?
R: Ele vai tentar fazer a validação com o valor de item.restriction_id correto? Mas isso está errado. Ele tem que tentar fazer a validação com null (O novo valor de restriction_id).

annecchini replied:
Não sei como fazer naquela versão reduzida, mas acredito que na forma abaixo as regras ficam mais claras do que naquele primeiro esboço que fizemos.
```
let region_id
if (typeof body.region_id !== 'undefined' && (body.region_id || body.region_id === null)) region_id = body.region_id
if (typeof body.region_id === 'undefined' && (item.region_id || item.region_id === null)) region_id = item.region_id
```

annecchini replied:
Considerando que:

Os valores de body.region_id já foram validados em etapa anterior (validateRegionId).
Item.region_id nunca será undefined (Eu sempre injeto na função de update).

 Daria para fazer menos código ainda.

```
const region_id = typeof body.region_id !== 'undefined' ? body.region_id : item.region_id
```
Assim podemos fazer const e com menos código.
Isso procede?

annecchini reviewed (changes requested):
Cara, não testei no postman, mas acredito que vai validar com o valor errado. Veja o comentário.

edo9k committed:
Added duplicate validation to update route

Also changed how the duplicate validation works, removed most mentions
to "item" variable, since it only used for sending an id when updating
an existing element (Vacancy).

edo9k committed:
Added duplicate validation to update route

Also changed how the duplicate validation works, removed mentions
to "item" variable in the duplicate validation, since it only used for sending an id when updating
an existing element (Vacancy). I might remove it in all other functions
where it is not needed before the final commit in this branch.

edo9k committed:
Merge branch 'feat/vacancy_duplicate_check' of https://github.com/SEAD-UFES/publications into feat/vacancy_duplicate_check

annecchini commented:
Mudanças efetuadas em outro branch. (fix/unique_Vacancy)
Fechando o pull request.
